src/transport/types.ts

```diff
@@ -0,0 +1,73 @@
+import events from 'events'
+import Multiaddr from 'multiaddr'
+import Connection from '../connection/connection'
+import { Sink } from '../stream-muxer/types'
+
+export type DialOptions = {
+  signal?: AbortSignal
+}
+
+/**
+ * A libp2p transport is understood as something that offers a dial and listen interface to establish connections.
+ */
+export interface Transport <DialOptions extends { signal?: AbortSignal }> {
+  new (upgrader: Upgrader, ...others: any): Transport<DialOptions>; // eslint-disable-line
+  prototype: Transport <DialOptions>;
+  /**
+   * Dial a given multiaddr.
+   */
+  dial(ma: Multiaddr, options?: DialOptions): Promise<Connection>;
+  /**
+   * Create transport listeners.
+   */
+  createListener(options: any, handler: (Connection) => void): Listener;
+  /**
+   * Takes a list of `Multiaddr`s and returns only valid addresses for the transport
+   */
+  filter(multiaddrs: Multiaddr[]): Multiaddr[];
+}
+
+export interface Listener extends events.EventEmitter {
+  /**
+   * Start a listener
+   */
+  listen(multiaddr: Multiaddr): Promise<void>;
+  /**
+   * Get listen addresses
+   */
+  getAddrs(): Multiaddr[];
+  /**
+   * Close listener
+   *
+   * @returns {Promise<void>}
+   */
+  close(): Promise<void>;
+}
+
+export interface Upgrader {
+  /**
+   * Upgrades an outbound connection on `transport.dial`.
+   */
+  upgradeOutbound(maConn: MultiaddrConnection): Promise<Connection>;
+
+  /**
+   * Upgrades an inbound connection on transport listener.
+   */
+  upgradeInbound(maConn: MultiaddrConnection): Promise<Connection>;
+}
+
+export type MultiaddrConnectionTimeline = {
+  open: number;
+  upgraded?: number;
+  close?: number;
+}
+
+export type MultiaddrConnection = {
+  sink: Sink;
+  source: () => AsyncIterable<Uint8Array>;
+  close: (err?: Error) => Promise<void>;
+  conn: any;
```

Comment:
I really don't like any here, I think it's best to type it or remove it or at least type as `conn: unknown` and make the consumer of the `MultiaddrConnection` check what is it rather than treat it as anything.